File: solutions/fsharp/phone-number/1/PhoneNumber.fs
module PhoneNumber

let private secondCheck input =
    if Seq.item 0 input < '2' then
        Error ("area code cannot start with " + if Seq.item 0 input = '1' then "one" else "zero")
    elif Seq.item 3 input < '2' then
        Error ("exchange code cannot start with " + if Seq.item 3 input = '1' then "one" else "zero")
    else 
        System.String(input |> Seq.toArray) |> System.UInt64.Parse |> Ok


let clean input =
    let onlyNumbers = Seq.filter System.Char.IsDigit input

    if Seq.exists System.Char.IsLetter input then
        Error "letters not permitted"
    elif Seq.exists (fun c -> set ['@'; ':'; '!'] |> Set.contains c) input then
        Error "punctuations not permitted"
    elif Seq.length onlyNumbers < 10 then
        Error "incorrect number of digits"
    elif Seq.length onlyNumbers = 11 && Seq.head onlyNumbers <> '1' then
        Error "11 digits must start with 1"
    elif Seq.length onlyNumbers > 11 then
        Error "more than 11 digits"
    elif Seq.length onlyNumbers = 11 then
        onlyNumbers |> Seq.tail |> secondCheck
    else
        onlyNumbers |> secondCheck

File: solutions/python/phone-number/1/phone_number.py
class PhoneNumber:
    __banned_punct = "@:!"
    __banned_area_exchange = "01"
    __country_code = 0
    __area_code = 0
    __exchange_code = 3

    def __init__(self, number):
        if any(c.isalpha() for c in number):
            raise ValueError("letters not permitted")
        if any(c in PhoneNumber.__banned_punct for c in number):
            raise ValueError("punctuations not permitted")

        number = "".join(c for c in number if c.isdigit())

        if len(number) < 10:
            raise ValueError("must not be fewer than 10 digits")
        if len(number) > 11:
            raise ValueError("must not be greater than 11 digits")

        if len(number) == 11:
            if number[PhoneNumber.__country_code] != "1":
                raise ValueError("11 digits must start with 1")
            number = number[PhoneNumber.__country_code + 1 :]

        if number[PhoneNumber.__area_code] in PhoneNumber.__banned_area_exchange:
            raise ValueError(
                "area code cannot start with "
                + ("one" if number[PhoneNumber.__area_code] == "1" else "zero")
            )
        if number[PhoneNumber.__exchange_code] in PhoneNumber.__banned_area_exchange:
            raise ValueError(
                "exchange code cannot start with "
                + ("one" if number[PhoneNumber.__exchange_code] == "1" else "zero")
            )

        self.number = number
        self.area_code = number[: PhoneNumber.__exchange_code]

    def pretty(self):
        return f"({self.number[:3]})-{self.number[3:6]}-{self.number[6:]}"
